### ISSUE TITLE
/boot/efi no mounted

### ISSUE THREAD

**Issue by minggaard** [saveriomiroddi/zfs-installer#140]
It seems as the sequence of /boot/efi and /boot in fstab can cause /boot/efi not to be mounted during boot.
Please consider changing the script so that /boot are included before adding entry for /boot/efi

**Comment by saveriomiroddi** [comment]
> It seems as the sequence of /boot/efi and /boot in fstab can cause /boot/efi not to be mounted during boot.
> Please consider changing the script so that /boot are included before adding entry for /boot/efi

Hello! Thanks for the report, it's very valuable. I'll check this within a week, then get back if I need further clarifications :smile: 

If you have some more details in the meanwhile, that'd be appreciated, however, without having a look, I don't know if further are meaningful :slightly_smiling_face:

**Comment by saveriomiroddi** [comment]
Hello!

So, I had a look today.

What I've observed in my testing environment is that, in `/etc/fstab`, `/boot/efi` is listed before `/boot`, however, in the syslog, the ordering is different:

```
$ sudo grep -P 'Mount.*/boot' /var/log/syslog
[...] Mounting /boot...
[...] Mounted /boot.
[...] Mounting /boot/efi2...
[...] Mounting /boot/efi...
[...] Mounted /boot/efi2.
[...] Mounted /boot/efi.
```

This implies that the mount is performed in parallel, therefore, changing the ordering won't make any difference.

I'm not experienced in this area. It seems that a mount dependency should be introduced.

It'd be best if you wrote directly to them [here](https://zfsonlinux.topicbox.com/groups/zfs-discuss), for a few reasons - you can reproduce the problem (I can't); they're move experienced; the installer is based on the [official document](https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2018.04%20Root%20on%20ZFS.html), therefore, changes should be first addressed there.

The installer will be updated to the [official 20.04 procedure](https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2020.04%20Root%20on%20ZFS.html) in the future, which sidesteps the /boot mounting entirely (based on my understanding), and therefore solves this problem, however, I don't have the resources to update now.

**Comment by saveriomiroddi** [comment]
A thought - would you check out you `/var/log` when the issue happens? I wonder if the problem is caused by something else, rather than the mount order.

**Comment by koenvervloesem** [comment]
I noticed something that maybe is related. I see this in my `journalctl` output on Ubuntu 20.04:

```
Oct 18 09:31:31 mu systemd[1]: Finished zfs-import-bpool.service.
Oct 18 09:31:31 mu systemd[1]: dev-disk-by\x2dpartuuid-8aac7233\x2dc9f5\x2d4390\x2d9c98\x2df6f6e053dbe3.device: Job dev-disk-by\x2dpartuuid-8aac7233\x2dc9f5\x2d4390\x2d9c98\x2df6f6e053dbe3.device/start timed out.
Oct 18 09:31:31 mu systemd[1]: Timed out waiting for device /dev/disk/by-partuuid/8aac7233-c9f5-4390-9c98-f6f6e053dbe3.
Oct 18 09:31:31 mu systemd[1]: Dependency failed for File System Check on /dev/disk/by-partuuid/8aac7233-c9f5-4390-9c98-f6f6e053dbe3.
Oct 18 09:31:31 mu systemd[1]: Dependency failed for /boot/efi.
Oct 18 09:31:31 mu systemd[1]: boot-efi.mount: Job boot-efi.mount/start failed with result 'dependency'.
Oct 18 09:31:31 mu systemd[1]: systemd-fsck@dev-disk-by\x2dpartuuid-8aac7233\x2dc9f5\x2d4390\x2d9c98\x2df6f6e053dbe3.service: Job systemd-fsck@dev-disk-by\x2dpartuuid-8aac7233\x2dc9f5\x2d4390\x2d9c98\x2df6f6e053dbe3.service/start failed with result 'dependency'.
Oct 18 09:31:31 mu systemd[1]: dev-disk-by\x2dpartuuid-8aac7233\x2dc9f5\x2d4390\x2d9c98\x2df6f6e053dbe3.device: Job dev-disk-by\x2dpartuuid-8aac7233\x2dc9f5\x2d4390\x2d9c98\x2df6f6e053dbe3.device/start failed with result 'timeout'.
```

My `/etc/fstab` looks like this:

```
koan@mu:~$ cat /etc/fstab 
PARTUUID=8aac7233-c9f5-4390-9c98-f6f6e053dbe3 /boot/efi vfat nofail,x-systemd.device-timeout=1 0 1
bpool /boot zfs nodev,relatime,x-systemd.requires=zfs-import-bpool.service 0 0
/dev/zvol/rpool/swap none swap discard 0 0
```

However, although I see this error message, `/boot/efi` apparently is still mounted later.

If I change the value of `x-systemd.device-timeout` to 5 in `/etc/fstab`, the error message disappears. Maybe that's something to try, @minggaard?

**Comment by minggaard** [comment]
It seems as the order does not matter. But there seems to be a timing issue as mentioned by koenvervloesem have the same in my systems too.
But I'll not trace this issue since the whole installer seems to be redone using the Ubuntu 20.4 at some point and I would rebuild then to get a bit more flexible disk layout.

**Comment by saveriomiroddi** [comment]
@koenvervloesem first, thanks for contributing to the discussion.

@minggaard is correct - there will be an update to the new reference procedure, which should radically solve this problem, although I'm in shortage of time, and I have no idea if/when it will ever happen (in particular, considering that this installer is a very niche program).

For general information, the new procedure will work on all the O/Ss. The official ZFS guide refers to it as 20.04, however, to be more precise, it's based on the 0.8 version of the ZFS libraries, which this installer installs on all the O/Ss :smile:

**Comment by saveriomiroddi** [comment]
@minggaard @koenvervloesem I've started the update work. Release 0.4.0, which is now written and under testing, will fix it (see #177).

I'm going to test it on a few O/Ss. If this feature is still useful for you, testing on systems other than mine would make the script more solid :smile: If not, I'm going to release it in stable once I'm done with testing on a few O/Ss.
